loadbalancer/zk/publisher_test.go

```diff
@@ -43,13 +44,16 @@ func TestBadFactory(t *testing.T) {
 	}
 	defer p.Stop()
 
+	// instance1 came online
+	client.AddService(path+"/instance1", "zookeeper_node_data")
+	time.Sleep(1 * time.Millisecond)
```

Comment:
I've used 2 to some success in the past. Another option, a signal chan:

```go
c := make(chan result, 1)
go func() {
    // do work
    c <- res
}()
select {
case res := <-c:
    if want, have := foo, res.value; want != have {
        t.Errorf("want %v, have %v", want, have)
    }
case <-time.After(100*time.Millisecond):
    t.Fatalf("timeout waiting for foo")
}
```

You can sometimes wrap that into a helper,

```
func within(d time.Duration, f func() error) error {
    c := make(chan error, 1)
    go func() { e <- f() }()
    select {
    case err := <-c:
        return err
    case <-time.After(d):
        return errors.New("timeout")
    }
}
```